File: apps/desktop/src/models/events.ts
/** Agent lifecycle states */
export type AgentStatus =
  | "Idle"
  | "Planning"
  | "WaitingPlanApproval"
  | "Thinking"
  | "WaitingApproval"
  | "RunningTool"
  | "Stopped"
  | "Completed"
  | "Failed";

/** A chain-of-thought thinking step emitted by the agent */
export interface ThinkingStep {
  id: string;
  content: string;
  timestamp: string;
}

/** A tool call proposed by the agent, awaiting user approval */
export interface ToolCallRequest {
  id: string;
  tool: string;
  params: Record<string, unknown>;
  /** Risk level hint from the agent */
  risk: "low" | "medium" | "high";
  /** Human-readable description of what the tool will do */
  description: string;
  timestamp: string;
}

/** Result after a tool call completes */
export interface ToolCallResult {
  id: string;
  tool: string;
  success: boolean;
  output?: unknown;
  error?: string;
  duration_ms: number;
  timestamp: string;
}

/** A chat message from user or agent */
export interface ChatMessage {
  id: string;
  role: "user" | "assistant" | "system";
  content: string;
  timestamp: string;
}

/** A plan step produced by the agent (for PlanUpdated — execution tracking) */
export interface PlanStep {
  id: string;
  index: number;
  description: string;
  status: "pending" | "running" | "done" | "failed" | "skipped";
}

/** An audit log entry */
export interface AuditEntry {
  id: string;
  session_id: string;
  event_type: string;
  payload: unknown;
  timestamp: string;
}

// ── RunGraph types ────────────────────────────────────────────────────────────

export type RunNodeKind = "tool_call" | "screenshot" | "verifier" | "user_message" | "agent_message";
export type RunNodeStatus = "pending" | "running" | "completed" | "failed" | "blocked";
export type EdgeKind = "control" | "data" | "verification";
export type ClaimResult = "pass" | "fail" | "skip";

export interface RunNode {
  id: string;
  session_id: string;
  kind: RunNodeKind;
  tool?: string;
  status: RunNodeStatus;
  confidence?: number;
  inputs?: unknown;
  outputs?: unknown;
  timestamp: string;
}

export interface RunEdge {
  id: string;
  session_id: string;
  src: string;
  dst: string;
  kind: EdgeKind;
  blocked: boolean;
  timestamp: string;
}

export interface VerifierClaim {
  id: string;
  session_id: string;
  run_node_id: string;
  verifier: string;
  result: ClaimResult;
  score?: number;
  detail?: string;
  timestamp: string;
}

export interface Intervention {
  id: string;
  session_id: string;
  kind: "block_edge" | "replace_artifact" | "insert_verifier";
  payload: unknown;
  timestamp: string;
/** A step in an AgentPlan produced by the Planner */
export interface AgentPlanStep {
  id: string;
  index: number;
  intent: string;
  tool: string;
  params_preview: Record<string, unknown>;
  depends_on: string[];
  risk: "low" | "medium" | "high";
  needs_credential?: string;
}

/** A structured plan produced by the Planner */
export interface AgentPlan {
  id: string;
  goal: string;
  steps: AgentPlanStep[];
  overall_risk: "low" | "medium" | "high";
  requires_credentials: string[];
}

/** Union of all events emitted by the backend */
export type AgentEvent =
  | { type: "StatusChanged"; status: AgentStatus }
  | { type: "ChatMessage"; message: ChatMessage }
  | { type: "PlanUpdated"; steps: PlanStep[] }
  | { type: "PlanningStarted" }
  | { type: "PlanReady"; plan: AgentPlan }
  | { type: "ToolCallRequest"; request: ToolCallRequest }
  | { type: "ToolCallResult"; result: ToolCallResult }
  | { type: "AuditEntry"; entry: AuditEntry }
  | { type: "AgentThinking"; step: ThinkingStep }
  | { type: "EmergencyStop" }
  // RunGraph events
  | { type: "GraphNodeAdded"; node: RunNode }
  | { type: "GraphNodeUpdated"; node: RunNode }
  | { type: "GraphEdgeAdded"; edge: RunEdge }
  | { type: "ArtifactCreated"; artifact_id: string; run_node_id: string; kind: string }
  | { type: "VerifierResult"; claim: VerifierClaim }
  | { type: "InterventionApplied"; intervention: Intervention };

File: apps/desktop/src-tauri/src/events.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// All agent lifecycle states.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum AgentStatus {
    Idle,
    Thinking,
    WaitingApproval,
    RunningTool,
    Stopped,
    Completed,
    Failed,
}

/// Risk level of a proposed tool call.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RiskLevel {
    Low,
    Medium,
    High,
}

/// A tool call proposed by the agent, awaiting user approval.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCallRequest {
    pub id: String,
    pub tool: String,
    pub params: serde_json::Value,
    pub risk: RiskLevel,
    pub description: String,
    pub timestamp: DateTime<Utc>,
}

/// Result after a tool call completes.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCallResult {
    pub id: String,
    pub tool: String,
    pub success: bool,
    pub output: Option<serde_json::Value>,
    pub error: Option<String>,
    pub duration_ms: u64,
    pub timestamp: DateTime<Utc>,
}

/// A chat message from user or agent.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatMessage {
    pub id: String,
    pub role: String, // "user" | "assistant" | "system"
    pub content: String,
    pub timestamp: DateTime<Utc>,
}

/// A single plan step.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanStep {
    pub id: String,
    pub index: usize,
    pub description: String,
    pub status: PlanStepStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PlanStepStatus {
    Pending,
    Running,
    Done,
    Failed,
    Skipped,
}

/// Union of all events emitted to the frontend.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum AgentEvent {
    StatusChanged { status: AgentStatus },
    ChatMessage { message: ChatMessage },
    PlanUpdated { steps: Vec<PlanStep> },
    ToolCallRequest { request: ToolCallRequest },
    ToolCallResult { result: ToolCallResult },
    EmergencyStop,
}

impl AgentEvent {
    pub const EVENT_NAME: &'static str = "agent-event";
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // ── AgentStatus ───────────────────────────────────────────────────────────

    #[test]
    fn test_agent_status_variants_are_distinct() {
        assert_ne!(AgentStatus::Idle, AgentStatus::Thinking);
        assert_ne!(AgentStatus::Thinking, AgentStatus::WaitingApproval);
        assert_ne!(AgentStatus::WaitingApproval, AgentStatus::RunningTool);
        assert_ne!(AgentStatus::RunningTool, AgentStatus::Completed);
        assert_ne!(AgentStatus::Completed, AgentStatus::Failed);
        assert_ne!(AgentStatus::Failed, AgentStatus::Stopped);
    }

    #[test]
    fn test_agent_status_clone_equals_original() {
        let s = AgentStatus::RunningTool;
        assert_eq!(s.clone(), AgentStatus::RunningTool);
    }

    #[test]
    fn test_agent_status_serializes_to_string() {
        let json = serde_json::to_value(AgentStatus::Thinking).unwrap();
        assert_eq!(json.as_str().unwrap(), "Thinking");
        let json = serde_json::to_value(AgentStatus::WaitingApproval).unwrap();
        assert_eq!(json.as_str().unwrap(), "WaitingApproval");
    }

    #[test]
    fn test_agent_status_round_trips_through_json() {
        let statuses = [
            AgentStatus::Idle,
            AgentStatus::Thinking,
            AgentStatus::WaitingApproval,
            AgentStatus::RunningTool,
            AgentStatus::Stopped,
            AgentStatus::Completed,
            AgentStatus::Failed,
        ];
        for s in &statuses {
            let json = serde_json::to_value(s).unwrap();
            let back: AgentStatus = serde_json::from_value(json).unwrap();
            assert_eq!(back, *s);
        }
    }

    // ── AgentEvent ────────────────────────────────────────────────────────────

    #[test]
    fn test_event_name_constant() {
        assert_eq!(AgentEvent::EVENT_NAME, "agent-event");
    }

    #[test]
    fn test_status_changed_event_serializes_with_type_tag() {
        let event = AgentEvent::StatusChanged {
            status: AgentStatus::Idle,
        };
        let json = serde_json::to_value(&event).unwrap();
        assert_eq!(json["type"], "StatusChanged");
        assert_eq!(json["status"], "Idle");
    }

    #[test]
    fn test_emergency_stop_event_serializes_with_type_tag() {
        let event = AgentEvent::EmergencyStop;
        let json = serde_json::to_value(&event).unwrap();
        assert_eq!(json["type"], "EmergencyStop");
    }

    #[test]
    fn test_chat_message_event_contains_message_field() {
        let msg = ChatMessage {
            id: "msg-1".into(),
            role: "user".into(),
            content: "hello".into(),
            timestamp: chrono::Utc::now(),
        };
        let event = AgentEvent::ChatMessage { message: msg };
        let json = serde_json::to_value(&event).unwrap();
        assert_eq!(json["type"], "ChatMessage");
        assert_eq!(json["message"]["role"], "user");
        assert_eq!(json["message"]["content"], "hello");
    }

    // ── RiskLevel ─────────────────────────────────────────────────────────────

    #[test]
    fn test_risk_level_serializes_lowercase() {
        assert_eq!(
            serde_json::to_value(RiskLevel::Low).unwrap().as_str().unwrap(),
            "low"
        );
        assert_eq!(
            serde_json::to_value(RiskLevel::Medium).unwrap().as_str().unwrap(),
            "medium"
        );
        assert_eq!(
            serde_json::to_value(RiskLevel::High).unwrap().as_str().unwrap(),
            "high"
        );
    }

    // ── PlanStepStatus ────────────────────────────────────────────────────────

    #[test]
    fn test_plan_step_status_serializes_lowercase() {
        assert_eq!(
            serde_json::to_value(PlanStepStatus::Pending).unwrap().as_str().unwrap(),
            "pending"
        );
        assert_eq!(
            serde_json::to_value(PlanStepStatus::Done).unwrap().as_str().unwrap(),
            "done"
        );
        assert_eq!(
            serde_json::to_value(PlanStepStatus::Failed).unwrap().as_str().unwrap(),
            "failed"
        );
        assert_eq!(
            serde_json::to_value(PlanStepStatus::Skipped).unwrap().as_str().unwrap(),
            "skipped"
        );
    }
}
